semgrep-core/src/engine/Match_rules.ml

```diff
@@ -938,21 +939,29 @@ and lazy_ast_of_string lang base_tok str warn_msg =
          if errors <> [] then pr2 (warn_msg ());
          (ast, errors)))
 
-and eval_nested_formula env formula lazy_ast_and_errors lazy_content opt_context
-    =
-  (* the following code is very similar to `check' below;
-     * we could maybe factorize things *)
+and nested_formula_has_matches env formula lazy_ast_and_errors lazy_content
+    opt_context =
+  let _, final_ranges =
+    evaluate_formula env formula lazy_ast_and_errors lazy_content opt_context
+  in
+  match final_ranges with [] -> false | _ :: _ -> true
+
+and evaluate_formula env formula lazy_ast_and_errors lazy_content opt_context =
```

Comment:
Nice refactoring here, minor comment: I would perhaps move this declaration up closer to `compose_patterns`.